jupyterhub/spawner.py

```diff
@@ -182,7 +182,74 @@ def _deprecate_percent_u(self, proposal):
         from having an effect on their server.
         """
     ).tag(config=True)
-    
+
+    mem_limit = MemorySpecification(
+        None,
+        allow_none=True,
+        help="""
+        Maximum number of bytes a single-user server is allowed to use.
+
+        Allows the following suffixes:
+          - K -> Kilobytes
+          - M -> Megabytes
+          - G -> Gigabytes
+          - T -> Terabytes
+
+        If the single user server tries to allocate more memory than this,
+        it will fail. There is no guarantee that the single-user server
+        will be able to allocate this much memory - only that it can not
+        allocate more than this.
+
+        This needs to be supported by your spawner for it to work.
+        """
+    ).tag(config=True)
+
+    cpu_limit = Float(
+        None,
```

Comment:
For consistency, mainly, I like keeping a typed value to always have its type. But I'm okay with None here if you prefer.